[DTensor] have split_strategy return OpStrategy instead of TupleStrategy

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158112
* __->__ #158051

**Summary**
`split_strategy` used `TupleStrategy` as return type because DTensor sharding
propagation's `OpStrategy` support on multi-returns only applies to `Tuple`.

However, `TupleStrategy`'s not a good fit for `split` op. `TupleStrategy` was
initially introduced to handle the sharding strategy of `foreach_*` ops where
the input args can be split into independent subsets regarding sharding decisions,
so are the outputs.

To address the misuse, this PR adds `OpStrategy` propagation for `List[Tensor]`
(note that this support is INCOMPLETE because it only checks the return type
to be `torch.ListType`). Nevertheless, the logic for `Tuple` returns also made
similar assumption so I think it's fine to unblock in such a way.

Besides adding `OpStrategy` support to ops having `List[Tensor]` return type,
this PR also changes `split_strategy`'s return from `TupleStrategy` to `OpStrategy`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

XilunWu committed:
[DTensor] have split_strategy return OpStrategy instead of TupleStrategy

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -353,6 +353,8 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
                                 for _ in range(len(op_schema.op._schema.returns))
                             ]
                         )
+                elif op_schema.return_type_list_tensor_like():
```
i will check if the above changes to split strategy works with AP.

XilunWu replied:
good catch. There're two types of ops:
1. return a Tuple of torch.Tensor.
2. return a List of torch.Tensor.

The main difference in OpStrategy is its `output_specs`. For (1), the output_specs is a `DTensorSpec` which in sharding prop we expand it to a List of repeated `DTensorSpec`. For (2), the output_specs is already the List of repeated `DTensorSpec`.

If this troubles you, I can modify the behavior to have (2) align with (1).

zpcore commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -1003,23 +1003,27 @@ def size_split(N, i) -> list:
+            # if the input is sharded on the split dim, we need to unshard it
+            placements = unshard_tensor_dim(spec.placements, dim=dim)
+
+        input_spec = DTensorSpec(spec.mesh, placements)  # TODO: fix tensor_meta
```
We can use `tensor_meta=spec.tensor_meta`.

zpcore commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -1003,23 +1003,27 @@ def size_split(N, i) -> list:
+        input_spec = DTensorSpec(spec.mesh, placements)  # TODO: fix tensor_meta
+        output_specs = tuple(
+            DTensorSpec(spec.mesh, placements)
+            for _ in range(len(output_size_list))  # TODO: fix tensor_meta
```
The `tensor_meta` here for output_specs is not necessary.

wanchaol commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -353,6 +353,8 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
                                 for _ in range(len(op_schema.op._schema.returns))
                             ]
                         )
+                elif op_schema.return_type_list_tensor_like():
```
it should at least be folded into the next `elif` branch instead of making it separate

XilunWu commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -333,6 +333,13 @@ def return_type_tuple_tensor_like(self) -> bool:
+        # all dispatch ops could only return List[Tensor]
+        return_types = self.op._schema.returns
+        return len(return_types) == 1 and isinstance(
+            return_types[0].type, torch.ListType
```
right

XilunWu committed:
Update on "[DTensor] have split_strategy return OpStrategy instead of TupleStrategy"


**Summary**
`split_strategy` used `TupleStrategy` as return type because DTensor sharding
propagation's `OpStrategy` support on multi-returns only applies to `Tuple`.

However, `TupleStrategy`'s design is complicated and we want to avoid its use
as much as possible. This PR adds `OpStrategy` propagation for `List[Tensor]`
(note that this support is INCOMPLETE because it only checks the return type
to be `torch.ListType`). Nevertheless, the logic for `Tuple` returns also made
similar assumption so I think it's fine to unblock in such a way.

Besides adding `OpStrategy` support to ops having `List[Tensor]` return type,
this PR also changes `split_strategy`'s return from `TupleStrategy` to `OpStrategy`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -333,6 +333,13 @@ def return_type_tuple_tensor_like(self) -> bool:
         )
 
+    def return_type_list_tensor_like(self) -> bool:
+        # all dispatch ops could only return List[Tensor]
```
i think the confusing part (to me at least) is the grammar.  It sounds like a statement that all the dispatch ops in the world could only return List[Tensor].  I think what you're trying to say is "Applies to any dispatch ops that return only List[Tensor]" ?

XilunWu committed:
Update on "[DTensor] have split_strategy return OpStrategy instead of TupleStrategy"


**Summary**
`split_strategy` used `TupleStrategy` as return type because DTensor sharding
propagation's `OpStrategy` support on multi-returns only applies to `Tuple`.

However, `TupleStrategy`'s design is complicated and we want to avoid its use
as much as possible. This PR adds `OpStrategy` propagation for `List[Tensor]`
(note that this support is INCOMPLETE because it only checks the return type
to be `torch.ListType`). Nevertheless, the logic for `Tuple` returns also made
similar assumption so I think it's fine to unblock in such a way.

Besides adding `OpStrategy` support to ops having `List[Tensor]` return type,
this PR also changes `split_strategy`'s return from `TupleStrategy` to `OpStrategy`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

wconstab reviewed (approved):
thanks! i confirmed this works for me locally
